DOC: Remove references to PyCObject (#17154)
numpy/numpy: doc/source/reference/arrays.interface.rst

1. The PyArrayInterface structure had no descr member at the end
   (and therefore no flag ARR_HAS_DESCR)

2. The ``context`` member of the :c:type:`PyCapsule` (formally the ``desc``
   member of the ``PyCObject``) returned from ``__array_struct__`` was
   not specified.  Usually, it was the object exposing the array (so
   that a reference to it could be kept and destroyed when the
   C-object was destroyed). It is now an explicit requirement that this field
   be used in some way to hold a reference to the owning object.

   .. note::

       Until August 2020, this said:

           Now it must be a tuple whose first element is a string with "PyArrayInterface Version #" and whose
           second element is the object exposing the array.

       This design was retracted in <https://mail.python.org/pipermail/numpy-discussion/2006-June/020995.html>, and is an error to have relied

>        upon in the intermediate 14 years.

Reviewer: "in" <- "it"

Reviewer: Should also break the line.

Reviewer: Maybe "and its continued use has been an error since that decision."

Author: I think that decision was almost immediately after this document was published. Happy to reword this however

Reviewer: > Happy to reword this however

I found it a bit confusing, so a rewrite would help.

Author: Committed my suggestion below, let me know if that's any better.

>        This design was retracted in <https://mail.python.org/pipermail/numpy-discussion/2006-June/020995.html>, and is an error to have relied
>        upon in the intermediate 14 years.

Author: ```suggestion
       This design was retracted in
       <https://mail.python.org/pipermail/numpy-discussion/2006-June/020995.html>,
       and it is an error to have relied upon in the intermediate 14 years.
```

>        This design was retracted in <https://mail.python.org/pipermail/numpy-discussion/2006-June/020995.html>, and is an error to have relied
>        upon in the intermediate 14 years.

Author: How's this @charris?
```suggestion
       This design was retracted almost immediately after it was proposed, in
       <https://mail.python.org/pipermail/numpy-discussion/2006-June/020995.html>.
       Despite 14 years of documentation to the contrary, at no point was it valid
       to assume that ``__array_interface__`` capsules held this tuple content.
```

1. The PyArrayInterface structure had no descr member at the end
   (and therefore no flag ARR_HAS_DESCR)

2. The ``context`` member of the :c:type:`PyCapsule` (formally the ``desc``
   member of the ``PyCObject``) returned from ``__array_struct__`` was
   not specified.  Usually, it was the object exposing the array (so
   that a reference to it could be kept and destroyed when the
   C-object was destroyed). It is now an explicit requirement that this field
   be used in some way to hold a reference to the owning object.

>    .. note::
> 
>        Until August 2020, this said:
> 
>            Now it must be a tuple whose first element is a string with "PyArrayInterface Version #" and whose
>            second element is the object exposing the array.
> 
>        This design was retracted almost immediately after it was proposed, in
>        <https://mail.python.org/pipermail/numpy-discussion/2006-June/020995.html>.
>        Despite 14 years of documentation to the contrary, at no point was it valid
>        to assume that ``__array_interface__`` capsules held this tuple content.

Author: ```suggestion
   .. note::

       Until August 2020, this said:

           Now it must be a tuple whose first element is a string with
           "PyArrayInterface Version #" and whose second element is the object
           exposing the array.

       This design was retracted almost immediately after it was proposed, in
       <https://mail.python.org/pipermail/numpy-discussion/2006-June/020995.html>.
       Despite 14 years of documentation to the contrary, at no point was it
       valid to assume that ``__array_interface__`` capsules held this tuple
       content.
```

1. The PyArrayInterface structure had no descr member at the end
   (and therefore no flag ARR_HAS_DESCR)

2. The ``context`` member of the :c:type:`PyCapsule` (formally the ``desc``
   member of the ``PyCObject``) returned from ``__array_struct__`` was
   not specified.  Usually, it was the object exposing the array (so
   that a reference to it could be kept and destroyed when the
   C-object was destroyed). It is now an explicit requirement that this field
   be used in some way to hold a reference to the owning object.

>    .. note::
> 
>        Until August 2020, this said:
> 
>            Now it must be a tuple whose first element is a string with
>            "PyArrayInterface Version #" and whose second element is the object
>            exposing the array.
> 
>        This design was retracted almost immediately after it was proposed, in
>        <https://mail.python.org/pipermail/numpy-discussion/2006-June/020995.html>.
>        Despite 14 years of documentation to the contrary, at no point was it
>        valid to assume that ``__array_interface__`` capsules held this tuple
>        content.
> 
> 3. The tuple returned from ``__array_interface__['data']`` used to be a
>    hex-string (now it is an integer or a long integer).

Reviewer 2: ```suggestion
3. The tuple returned from ``__array_interface__['data']`` used to be a
   hex-string (now it is an integer or a long integer).
   .. note::

       Until August 2020, this said:

           Now it must be a tuple whose first element is a string with
           "PyArrayInterface Version #" and whose second element is the object
           exposing the array.

       This design was retracted almost immediately after it was proposed, in
       <https://mail.python.org/pipermail/numpy-discussion/2006-June/020995.html>.
       Despite 14 years of documentation to the contrary, at no point was it
       valid to assume that ``__array_interface__`` capsules held this tuple
       content.

```

Reviewer 2: Ahh, right. The "this" got me confused. So maybe "Until August 2020, the description of `context` said...", or just forget it and what you have is good.
